validator/client/src/main/java/tech/pegasys/teku/validator/client/ForkProvider.java

```diff
@@ -43,49 +43,63 @@ public ForkProvider(
     this.asyncRunner = asyncRunner;
     this.validatorApiChannel = validatorApiChannel;
     this.genesisDataProvider = genesisDataProvider;
+    currentFork = new SafeFuture<>();
   }
 
   public SafeFuture<ForkInfo> getForkInfo() {
-    return currentFork.map(SafeFuture::completedFuture).orElseGet(this::loadForkInfo);
+    return currentFork;
   }
 
-  public SafeFuture<ForkInfo> loadForkInfo() {
-    return requestForkInfo()
-        .exceptionallyCompose(
-            error -> {
-              LOG.error("Failed to retrieve current fork info. Retrying after delay", error);
-              return asyncRunner.runAfterDelay(
-                  this::getForkInfo, FORK_RETRY_DELAY_SECONDS, TimeUnit.SECONDS);
-            });
-  }
-
-  private SafeFuture<ForkInfo> requestForkInfo() {
-    return genesisDataProvider.getGenesisValidatorsRoot().thenCompose(this::requestFork);
-  }
-
-  public SafeFuture<ForkInfo> requestFork(final Bytes32 genesisValidatorsRoot) {
+  private SafeFuture<ForkInfo> getForkAndPeriodicallyUpdate() {
     return validatorApiChannel
         .getFork()
         .thenCompose(
             maybeFork -> {
               if (maybeFork.isEmpty()) {
                 LOG.trace("Fork info not available, retrying");
-                return asyncRunner.runAfterDelay(
-                    this::requestForkInfo, FORK_RETRY_DELAY_SECONDS, TimeUnit.SECONDS);
+                return scheduleNextRequest(FORK_RETRY_DELAY_SECONDS);
```

Comment:
This isn't right.  It will wind up recursively chaining the `exceptionallyCompose` at the end of this method until you potentially get a `StackOverflowException`.

The other places are ok to call `loadForkInfo` recursively because its either already in the last step of the chain or doesn't return the result and just uses `reportExceptions`.  That's why there was a split between `loadForkInfo` and `requestForkInfo`.